AMQCPP-756: Changes for C++17 compliance

arayq2 opened the pull request:
Gathering the issues in several related JIRAs into one PR.  The changes for C++17 compliance:

1. Replace std::auto_ptr with std::unique_ptr (with 6 fix-ups besides global SAR with a Perl one-liner).
2. Eliminate dynamic exception specifications (no longer supported in C++17).
3. Eliminate warnings when compiling with g++ (GCC 11.4.0).

Code base compiles cleanly, and all tests (regular and integration) pass.  

Environment:
Linux Ubuntu 22.04
GCC 11.4.0
OpenSSL 3.0.2

arayq2 committed:
NO-JIRA Eliminate NULL initializer of auto_ptr members.

    Not needed per se, and will interfere with replacement by unique_ptr.

arayq2 committed:
https://issues.apache.org/jira/browse/AMQCPP-664

AMQCPP-664 Global replace auto_ptr with unique_ptr; fix one test.

    std::auto_ptr has been deprecated since C++11 and removed in C++17.
    Fix one test: added an include (to resolve forward declaration).

arayq2 committed:
https://issues.apache.org/jira/browse/AMQCPP-718

    AMQCPP-718 Removed dynamic exception specification.

    throw(...) deprecated in C++11, removed in C++17.
    throw() retained as alias of noexcept until C++20.

arayq2 committed:
https://issues.apache.org/jira/browse/AMQCPP-756

    AMQCPP-756: Fixes for C++17 compiler (GCC 11.4.0) warnings

    Regular tests passed (2070 tests).
    Integration check passed (133 tests).

arayq2 committed:
NO-JIRA Administrivia.

    Update README.txt files - should be combined into a single README.md.
        How to deal with missing cppunit-config program.
        Example shell script added to distribution.
        Recommend building outside the source tree.

    Add directories to .gitignore file.
        .vscode/ - local IDE customizations.
        .deps/   - artifacts of GCC build.

mattrpav commented:
@arayq2 Thank you for the contribution! I know the importance of C++ and really appreciate your efforts to modernize the ActiveMQ C++ client library.

While my C++ expertise is very dated, I'm happy to help with testing, reviewing and supporting your effots.

Question: Are the steps to build and test in the README.md current? That is, if I checkout these changes, run the build and tests using that as the guide I should see all tests pass?

edit: Once I get the builds steps confirmed, I'll add Jenkins build and/or a GH action so we can start adding some modern CI/CD actions.

Thanks!
Matt Pavlovich
ActiveMQ PMC Member

arayq2 commented:
@mattrpav, thanks for your attention. You raise a good point.  The README.md is probably out of date for modern systems, specifically on configuring the cppunit dependency for the tests.  I should probably update it with my workaround. That will mean a commit added to the PR.  Stand by!

arayq2 commented:
Here is my cheesy shell script that I used to fake out the m4 macros in the build system.  The better approach, of course, would be to update the m4 macros to use pkg-config or equivalent so that this workaround isn't necessary.

[cppunit-config.txt](https://github.com/user-attachments/files/19039389/cppunit-config.txt)

It isn't clear whether it should be included in the package, as it's a workaround, and values are hardcoded.

artnaseef commented:
I have some time this weekend to look at this.  I just tried a first attempt at a build on my 20.04 Ubuntu system with gcc version 9.4.0 and got the following complaint:

```
./decaf/lang/Throwable.h:49:28: error: expected ‘;’ at end of member declaration
                           49 |         virtual ~Throwable() noexcept = default;
```

So I was wondering if this is an issue with the C++ standard applied.  Checking the actual compile line, I don't see a standard being chosen:

```
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I../.. -ansi -pedantic -DLINUX -D_REENTRANT -D_GNU_SOURCE -I/usr/include/apr-1.0 -W -Wall -Wextra -Wconversion -fPIC -fstrict-aliasing -Wstrict-aliasing=2 -Wno-long-long -g -O2 -pthread -MT activemq/cmsutil/libactivemq_cpp_la-CmsAccessor.lo -MD -MP -MF activemq/cmsutil/.deps/libactivemq_cpp_la-CmsAccessor.Tpo -c activemq/cmsutil/CmsAccessor.cpp  -fPIC -DPIC -o activemq/cmsutil/.libs/libactivemq_cpp_la-CmsAccessor.o
```

Can you help me track this down?  Does the `configure` script check which C++ standard to apply?  Or does the newer version of `g++` assume a different default standard?

artnaseef commented:
Update: found that adding `-std=c++17` and removing `-ansi` from the `g++` command works.  I'm going to dig through the configure scripts to see if I can figure out how `-ansi` is chosen.

artnaseef commented:
Looks like this is chosing `-ansi`:

`configure.ac:   PLAT_CXXFLAGS="-ansi -pedantic"`

Not finding any place that would be changed or overriden.

arayq2 commented:
I just set CXXFLAGS from the command line in the terminal. It can also be set as argument to the configure script, but if you're using build.sh, then the command line is better.
$ export CXXFLAGS="-std=c++17" 
or, in case gdb is needed later
$ export CXXFLAGS="-std=c++17 -g"
 
(I have it set by my .bashrc automatically when I start the terminal.  I guess I forgot about it. Sorry!)

artnaseef commented:
I am trying with the configure script - appears to be working.  The `build.sh` script is confusing as the README.txt's instructions just say:

> The easiest
way to enable this is to edit the build.conf file before running build.sh,
which will create an appropriate configure script and run it.  Then, simply
'cd build' before running make or the tests.

This is bringing back memories.  Long lost memories ;-).  Haven't had to deal with complicated instructions like this in a long time...  Can we update the `configure.ac` like this?

*OLD*
```
if test "$GCC" = "yes"; then
   PLAT_CXXFLAGS="-ansi -pedantic"
fi
```

*NEW*
```
if test "$GCC" = "yes"; then
   PLAT_CXXFLAGS="-std=c++17"
fi
```

My hope is we can get as close as possible to `./configure` then `make` just working out of the box.

arayq2 commented:
It may be better to leave PLAT_CXXFLAGS empty for GCC, and give configure the correct CXXFLAGS depending on local context.  E.g. export CXXFLAGS="-std=c++11" (I did that just for kicks and everything compiled!!)  The "-ansi" setting is _ancient_ (definitely pre-C++11) and hasn't been used in years. 

I just checked my build logs, and it looks like -std=c++17 will cause -ansi to be ignored.  configure picks up both PLAT_CXXFLAGS and CXXFLAGS, though why there is an extra test outside the case statement in configure.ac is a mystery.  Looks like an ad hoc hack inserted later.

artnaseef commented:
How can we make this as easy as possible for users of this project, at least when using a compatible / standard environment?

BTW, I tried editing the `g++` command by hand and found that `-std=c++17` does not work if it comes before `-ansi` but works if it comes after.  So really, all of this type of complexity sucks for our users - hence why I would like `configure.ac` to figure out the right answer out-of-the-box.  Thinking "convention-over-configuration" here - out of the box, if you are buildig on a compatible environment, `./configure` followed by `make` then `make install` just works.

What do you think?

artnaseef commented:
I just tested with `-ansi -pedantic` removed (single file compile) and it appears to work, so just emptying `PLAT_CXXFLAGS` works for me.

arayq2 commented:
I've been looking at these various config and build related files, and I agree that they need work.  For example, I've checked the two README files and determined that the one at the distro root is obsolete. We should probably nuke it lest it cause confusion, and convert the other one to .md format.

There's stuff in configure.ac that looks iffy, such as the PLAT_CXXFLAGS setting for GCC outside the case statement. (Does that mean that GCC is not used at all on those platforms?) I think it should be removed as well.

And then there is the business of cppunit-config being obsolete and missing on modern systems.  pkg-config is neither a drop-in replacement, nor is it universal (I believe it is Debian/Ubuntu specific).  My shell script is a terrible hack but right now I can't think of a good way to automate that part.

arayq2 commented:
I'm not sure that setting CXXFLAGS inside configure.ac is the right way to go.  The code base compiles cleanly with C++11.  The issue I fixed was that compilation was failing for C++17; now it doesn't.  There is no need to _insist_ on C++17.  In fact, as far as I know, C++20 would fail only because 'throw()' is no longer recognized as an alias for 'noexcept'. A global SAR one-liner can fix that as well.
 Also, the GCC docs state that C++17 is the default setting for GCC 11, so not setting -std at all will compile under C++17 - it's the -ansi setting that's screwing things up.  Just let people use whatever version of GCC they have.

artnaseef committed:
Remove "-ansi -pedantic" as default GCC options.  Work-around Long.cpp bug when optimized.  Date test less fragile - force expected TZ used by test.

- The "-ansi -pedantic" options force a really old standard of C/C++ to be used
- The Long.cpp bug causes overflow detection to fail, leading to missing NumberFormatExceptions
- The Date test was always timezone-specific, making it fragile to run anywhere outside that timezone (EDT)

artnaseef committed:
Remove "-ansi -pedantic" as default GCC options.  Work-around Long.cpp bug when optimized.  Date test less fragile - force expected TZ used by test.

- The "-ansi -pedantic" options force a really old standard of C/C++ to be used
- The Long.cpp bug causes overflow detection to fail, leading to missing NumberFormatExceptions
- The Date test was always timezone-specific, making it fragile to run anywhere outside that timezone (EDT)

artnaseef committed:
Use pkg-config for cppunit on modern systems, if cppunit-config is not found.

arayq2 committed:
Restore the logic for the unlikely event that cppunit-config is found.

arayq2 committed:
Merge branch 'artnaseef-asn-fixup-2025-03' into fix/AMQCPP-756

arayq2 committed:
NO-JIRA Consolidated two README.txt files into a single README.md

arayq2 commented:
I've created 3 new JIRAs to cover the work here, esp by Art, that didn't have JIRA references

AMQCPP-757  cppunit-config program is obsolete and dropped
AMQCPP-758  Consolidate README.txt files
AMQCPP-759  src/main/decaf/lang/Long.cpp does not detect overflow properly

These should get closed out with the release.
Still missing: update for RELEASE_NOTES.txt, which is doing the duty of a Change Log.

cquiroz commented:
Hello, I'm a bit confused on how this is being made public. is this change included on the 3.9.5 release?
